src/ucp/wireup/wireup_cm.c

```diff
@@ -41,6 +41,57 @@ int ucp_ep_init_flags_has_cm(unsigned ep_init_flags)
                                UCP_EP_INIT_CM_WIREUP_SERVER));
 }
 
+static const char* ucp_cm_used_idx_to_cm_name(ucp_context_h context,
+                                              ucp_rsc_index_t cms_used_idx,
+                                              char *buf, size_t max)
+{
+    ucp_rsc_index_t cmpt_index = context->config.cm_cmpt_idxs[cms_used_idx];
+
+    ucs_snprintf_safe(buf, max, "%s", context->tl_cmpts[cmpt_index].attr.name);
+    return buf;
+}
+
+/*
+ * The main thread progress part of attempting connecting the client to the server
+ * through the next available cm.
+ */
+static unsigned ucp_cm_client_try_next_cm_progress(void *arg)
+{
+    ucp_ep_h ucp_ep       = arg;
+    ucp_worker_h worker   = ucp_ep->worker;
+    ucp_context_h context = worker->context;
+    ucp_wireup_ep_t *cm_wireup_ep;
+    ucs_status_t status;
+    uct_ep_h uct_cm_ep;
+    char cm_name_prev[UCT_TL_NAME_MAX], cm_name_next[UCT_TL_NAME_MAX];
+
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    cm_wireup_ep = ucp_ep_get_cm_wireup_ep(ucp_ep);
+    ucs_assert_always(cm_wireup_ep != NULL);
+    uct_cm_ep = ucp_ep_get_cm_uct_ep(ucp_ep);
+    ucs_assert_always(uct_cm_ep != NULL);
+
+    /* try connecting through another cm (next one in the priority list) */
+    cm_wireup_ep->cms_used_idx++;
+    ucp_wireup_ep_unset_next_ep(cm_wireup_ep);
+    uct_ep_destroy(uct_cm_ep);
+
+    ucs_debug("client switching from %s to %s in attempt to connect to the server",
+              ucp_cm_used_idx_to_cm_name(context, cm_wireup_ep->cms_used_idx - 1, cm_name_prev, UCT_TL_NAME_MAX),
+              ucp_cm_used_idx_to_cm_name(context, cm_wireup_ep->cms_used_idx, cm_name_next, UCT_TL_NAME_MAX));
+
+    status = ucp_ep_client_cm_create_uct_ep(ucp_ep);
+    if (status != UCS_OK) {
+        ucs_error("failed to create a uct sockaddr endpoint on %s cm %p",
+                  ucp_cm_used_idx_to_cm_name(context, cm_wireup_ep->cms_used_idx, cm_name_next, UCT_TL_NAME_MAX),
+                  worker->cms[cm_wireup_ep->cms_used_idx].cm);
```

Comment:
maybe print error only if no one is OK?

Reply:
if creation of a cm ep fails, first one as well, i don't think we should fallback, it's a valid error to print to the user.
we fallback only on specific errors from the server side (in connect_cb) or if client_pack_cb fails.

Reply:
does it mean that in case of error there is a leak of UCP EP and wireup_ep?

Reply:
looks like it, so guess need to call ucp_worker_set_ep_failed() in case of failure to create a uct cm ep here

Reply:
IIRC `ucp_worker_set_ep_failed` will invoke user err handler, if we want to continue with next CM, we should not do this, need to clean up resources silently.

Reply:
won't the UCP EP be released from `ucp_worker_destroy` in this case?

Reply:
it will, but until that EPs will leak increasing resources consumption.